### ISSUE TITLE
Entity Framework 3  HttpException using Cosmos provider

### ISSUE THREAD

**Issue by ekjuanrejon** [aspnet/EntityFrameworkCore#15384]
Hello 

I am running .net core console application and I am constantly receiving the following error:
I am using EntityFramework core 3 preview

[09:55:57 ERR] An exception occurred while iterating over the results of a query for context type 'CareManagement.EventIntegration.Database.CareManagementDbContext'.
System.AggregateException: One or more errors occurred. (Only one usage of each socket address (protocol/network address/port) is normally permitted) ---> System.Net.Http.HttpRequestException: Only one usage of each socket address (protocol/network address/port) is normally permitted ---> System.Net.Sockets.SocketException: Only one usage of each socket address (protocol/network address/port) is normally permitted
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Azure.Cosmos.DocumentClient.HttpRequestMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.Azure.Cosmos.GatewayAccountReader.GetDatabaseAccountAsync(Uri serviceEndpoint)
   at Microsoft.Azure.Cosmos.Routing.GlobalEndpointManager.GetDatabaseAccountFromAnyLocationsAsync(Uri defaultEndpoint, IList`1 locations, Func`2 getDatabaseAccountFn)
   at Microsoft.Azure.Cosmos.GatewayAccountReader.InitializeReaderAsync()
   at Microsoft.Azure.Cosmos.Internal.CosmosAccountServiceConfiguration.InitializeAsync()
   at Microsoft.Azure.Cosmos.DocumentClient.InitializeGatewayConfigurationReader()
   at Microsoft.Azure.Cosmos.DocumentClient.GetInitializationTask()
   at Microsoft.Azure.Cosmos.DocumentClient.EnsureValidClientAsync()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Microsoft.Azure.Cosmos.DocumentClient.get_ConsistencyLevel()
   at Microsoft.Azure.Cosmos.CosmosClient.Init(CosmosConfiguration configuration, DocumentClient documentClient)
   at Microsoft.Azure.Cosmos.CosmosClient..ctor(CosmosConfiguration cosmosConfiguration)
   at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosClientWrapper.get_Client()
   at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosClientWrapper.CreateQuery(String containerId, CosmosSqlQuery query)
   at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosClientWrapper.DocumentAsyncEnumerable.AsyncEnumerator.MoveNext(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Expressions.Internal.QueryShaperExpression.AsyncShaperEnumerable`1.AsyncShaperEnumerator.MoveNext(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.FirstOrDefault_[TSource](IAsyncEnumerable`1 source, CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\First.cs:line 144
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.TaskResultAsyncEnumerable`1.Enumerator.MoveNext(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.SelectEnumerableAsyncIterator`2.MoveNextCore(CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\Select.cs:line 106
   at System.Linq.AsyncEnumerable.AsyncIterator`1.MoveNext(CancellationToken cancellationToken) in D:\a\1\s\Ix.NET\Source\System.Interactive.Async\AsyncIterator.cs:line 98
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteSingletonAsyncQuery[TResult](QueryContext queryContext, Func`2 compiledQuery, IDiagnosticsLogger`1 logger, Type contextType)
---> (Inner Exception #0) System.Net.Http.HttpRequestException: Only one usage of each socket address (protocol/network address/port) is normally permitted ---> System.Net.Sockets.SocketException: Only one usage of each socket address (protocol/network address/port) is normally permitted
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Azure.Cosmos.DocumentClient.HttpRequestMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.Azure.Cosmos.GatewayAccountReader.GetDatabaseAccountAsync(Uri serviceEndpoint)
   at Microsoft.Azure.Cosmos.Routing.GlobalEndpointManager.GetDatabaseAccountFromAnyLocationsAsync(Uri defaultEndpoint, IList`1 locations, Func`2 getDatabaseAccountFn)
   at Microsoft.Azure.Cosmos.GatewayAccountReader.InitializeReaderAsync()
   at Microsoft.Azure.Cosmos.Internal.CosmosAccountServiceConfiguration.InitializeAsync()
   at Microsoft.Azure.Cosmos.DocumentClient.InitializeGatewayConfigurationReader()
   at Microsoft.Azure.Cosmos.DocumentClient.GetInitializationTask()
   at Microsoft.Azure.Cosmos.DocumentClient.EnsureValidClientAsync()<---

**Comment by ajcvickers** [comment]
@ekjuanrejon What version of EF Core, .NET Core, and the Cosmos provider are you using?

**Comment by ekjuanrejon** [comment]
3.0.0-preview3.19153.1

**Comment by AndriySvyryd** [comment]
This is likely caused by not reusing the `CosmosClient` instance

**Comment by ekjuanrejon** [comment]
services
                .AddDbContext<CareManagementDbContext>(options =>
                {
                    options.UseCosmos(configuration["CosmosDb:EndpointUrl"], configuration["CosmosDb:PrivateKey"], configuration["CosmosDb:DbName"], optionBuilder =>
                    {
                        optionBuilder.ExecutionStrategy(d => new CosmosExecutionStrategy(d, 10));
                    });

                }, ServiceLifetime.Transient);

**Comment by ekjuanrejon** [comment]
If I don't use transient I get lots of errors about threads accessing the same dbcontext. I am running in a console application and not a web application. Must of the examples are based on web which use the scope

**Comment by AndriySvyryd** [comment]
@ekjuanrejon The bug is in the Cosmos provider where we don't scope `CosmosClient` correctly

**Comment by ekjuanrejon** [comment]
@AndriySvyryd When can it be fixed? We are looking to use cosmos db as part of our event sourcing but it is throwing loads of errors

**Comment by AndriySvyryd** [comment]
@ekjuanrejon  Not sooner than preview 6

**Comment by ekjuanrejon** [comment]
What's the time frame?

**Comment by AndriySvyryd** [comment]
We plan to ship a preview every month. See https://devblogs.microsoft.com/dotnet/announcing-net-core-3-preview-3/

**Comment by ekjuanrejon** [comment]
@AndriySvyryd what can be the work around. this is a high priority for us. Most of our stuff is ready to got but this is blocking us

**Comment by AndriySvyryd** [comment]
@ekjuanrejon Try adding this as the first call on the `ServiceCollection`:
```C#
.AddSingleton<Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosClientWrapper>()
```

You also might need to register the context as scoped and create a new scope each time.

**Comment by ekjuanrejon** [comment]
@AndriySvyryd when I do what you suggested. I get the following error

'A second operation started on this context before a previous operation completed. This is usually caused by different threads using the same instance of DbContext, however instance members are not guaranteed to be thread safe. This could also be caused by a nested query being evaluated on the client, if this is the case rewrite the query avoiding nested invocations.

**Comment by ekjuanrejon** [comment]
@AndriySvyryd 
when using async await it gives loads of null reference exception

**Comment by ekjuanrejon** [comment]
@AndriySvyryd 

it seems to work now so far.

Thanks

**Comment by ajcvickers** [comment]
@ekjuanrejon Just want to make sure you are aware that the Cosmos provider is in a pre-release state. This means that the quality is lower than a released version, it's not supported for production, and it will change in breaking ways before the RTM release happens.

**Comment by ekjuanrejon** [comment]
@ajcvickers thanks for feedback. When is the RTM release? Hopefully soon :)

**Comment by ajcvickers** [comment]
@ajcvickers It's currently planned to release with EF Core 3.0 later this year.

**Comment by AndriySvyryd** [comment]
Fixed in 98fe04febee96f733bbfd9c4e3c819a2cbe56a5b
